database.cc

```diff
@@ -846,9 +846,28 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
     }).then([&proxy, &mm, this] {
         return do_parse_schema_tables(proxy, db::schema_tables::VIEWS, [this, &proxy, &mm] (schema_result_value_type &v) {
             return create_views_from_schema_partition(proxy, v.second).then([this, &mm] (std::vector<view_ptr> views) {
-                return parallel_for_each(views.begin(), views.end(), [this, &mm] (auto&& v) {
-                    return this->add_column_family_and_make_directory(v).then([this, &mm, v] {
-                        return maybe_update_legacy_secondary_index_mv_schema(mm.local(), *this, v);
+                return do_with(std::vector<view_ptr>{}, [this, &mm, views = std::move(views)] (std::vector<view_ptr>& views_to_announce) {
+                    return parallel_for_each(views.begin(), views.end(), [this, &mm, &views_to_announce] (auto&& v) {
+                        // TODO: Remove once computed columns are guaranteed to be featured in the whole cluster.
+                        // we fix here the schema in place in oreder to avoid races (write commands comming from other coordinators).
+                        view_ptr fixed_v = maybe_fix_legacy_secondary_index_mv_schema(*this, v, nullptr, preserve_version::yes);
+                        // TODO: Remove once computed columns are guaranteed to be featured in the whole cluster.
+                        view_ptr v_to_add = v;
+                        if (bool(fixed_v)) {
+                            v_to_add = fixed_v;
+                            // now really fix the view by creating a new and updated version and registering it for announcement.
+                            views_to_announce.emplace_back(maybe_fix_legacy_secondary_index_mv_schema(*this, v, nullptr, preserve_version::no));
+                        }
+                        return this->add_column_family_and_make_directory(v_to_add);
+                    }).then([&views_to_announce, &mm, this] {
+                        if (!views_to_announce.empty()) {
+                            _computed_columns_reg = _feat.cluster_supports_computed_columns().when_enabled([&mm, views_to_announce = std::move(views_to_announce)] {
+                                (void)parallel_for_each(views_to_announce.begin(), views_to_announce.end(), [&mm] (const view_ptr& fixed_v) {
+                                    return mm.local().announce_view_update(view_ptr(fixed_v));
```

Comment:
Is it possible to have a race wrt. users modifying views during the upgrade process (not recommended, I know):

 1.   The cluster is only partially upgraded, so the computed columns feature is not yet enabled
  2.  A view with schema v1 is loaded and its schema is updated to v2, but not announced yet
  3.  A user alters view with schema v1 to v3
  4.  The upgrade finishes, we locally announce a local upgrade of this view to v2
  5.  We end up with an older version locally applied (?)

Or will the schemas be somehow merged and it would eventually work correctly?

Reply:
I believe that:
36d36269a67f2361c4c4195287b40bceeda1e65f from this series takes care of that,
Look in merge_chema function, we fix the view there in order to make it "write enabled" which means that the computed column stuff will be fine. About the question:
> Or will the schemas be somehow merged and it would eventually work correctly?

The whole idea of schema merge is to avoid clash with orthogonal changes so unless there is some weird or a very specific edge case it should be fine.

Reply:
I'm sure that it will be fine for computed columns, but I'm not sure about the schema in general (because I don't fully grasp how schema updates work), so I'll just post a very specific example and please correct me if it's impossible:
1. View v1 has size tiered compaction strategy
2. Schema v2 is prepared, it still has columns (a,b,c), but also correctly marks column c as computed
3. A user alters the view, creating schema v3, with leveled compaction strategy
4. The upgrade finishes and v2 is announced

Wouldn't the view end up with the original size tiered compaction strategy, even though a user set it to something else before the upgrade finished?

Reply:
I will need to look in the code for that. Will look into it and get back to you.

Reply:
At first look - there is a bug even more subtle than this. The race you mentioned doesn't exist because update view announce is differential. However, since we also fixing in place the diff will not catch the change at all because as far as it
is concerned the column is already marked as computed so the new view announcement will not contain the computed_columns mutations at all :facepalm: I will rework this.

Reply:
Good catch BTW :slightly_smiling_face:

Reply:
Last version should avoid the bug I mentioned in this thread.